pkg/csi/cinder/controllerserver.go

```diff
@@ -268,12 +316,51 @@ func (cs *controllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
 
 func (cs *controllerServer) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRequest) (*csi.ListSnapshotsResponse, error) {
 
-	filters := map[string]string{}
+	if len(req.GetSnapshotId()) != 0 {
+		snap, err := cs.Cloud.GetSnapshotByID(req.SnapshotId)
+
+		if err != nil {
+			return &csi.ListSnapshotsResponse{}, nil
+		}
+
+		ctime, err := ptypes.TimestampProto(snap.CreatedAt)
+
+		entry := &csi.ListSnapshotsResponse_Entry{
+			Snapshot: &csi.Snapshot{
+				SizeBytes:      int64(snap.Size * 1024 * 1024 * 1024),
+				SnapshotId:     snap.ID,
+				SourceVolumeId: snap.VolumeID,
+				CreationTime:   ctime,
+				ReadyToUse:     true,
+			},
+		}
+
+		entries := []*csi.ListSnapshotsResponse_Entry{entry}
+		return &csi.ListSnapshotsResponse{
+			Entries: entries,
+		}, err
+
+	}
+
 	// FIXME: honor the limit, offset and filters later
-	vlist, err := cs.Cloud.ListSnapshots(int(req.MaxEntries), 0, filters)
-	if err != nil {
-		klog.V(3).Infof("Failed to ListSnapshots: %v", err)
-		return nil, err
+	filters := map[string]string{}
+
+	var vlist []snapshots.Snapshot
+	var err error
+
+	if len(req.GetSourceVolumeId()) != 0 {
+		vlist, err = cs.Cloud.GetSnapshotByNameAndVolumeID("", req.GetSourceVolumeId())
+		if err != nil {
+			klog.V(3).Infof("Failed to ListSnapshots: %v", err)
+			return nil, err
```

Comment:
hmm, will check about it

Reply:
actually in case of no snapshot found to corresponding volumeID, empty list will be returned by GetSnapshotByNameAndVolumeID, for error case yes we need to return codes.internal